components/modal/index.js

```diff
@@ -0,0 +1,193 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+import { noop, uniqueId } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { Component, createPortal } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import ModalFrame from './frame';
+import ModalHeader from './header';
+import * as ariaHelper from './aria-helper';
+import './style.scss';
+
+// Used to count the number of open modals.
+let parentElement,
+	openModalCount = 0;
+
+class Modal extends Component {
+	constructor( props ) {
+		super( props );
+
+		this.prepareDOM();
+
+		this.setHeadingId( props );
+	}
+
+	/**
+	 * Prepares the DOM for the modals to be rendered.
+	 *
+	 * Every modal is mounted in a separate div appended to a parent div
+	 * that is appended to the document body.
+	 *
+	 * The parent div will be created if it does not yet exist, and the
+	 * separate div for this specific modal will be appended to that.
+	 */
+	prepareDOM() {
+		if ( ! parentElement ) {
+			parentElement = document.createElement( 'div' );
+			document.body.appendChild( parentElement );
+		}
+		this.node = document.createElement( 'div' );
+		parentElement.appendChild( this.node );
+	}
+
+	/**
+	 * Sets the heading id to the aria.labelledby prop or a unique id when
+	 * the prop is unavailable.
+	 *
+	 * @param {Object} props The component's props.
+	 */
+	setHeadingId( props ) {
+		this.headingId = props.aria.labelledby || uniqueId( 'modal-heading-' );
+	}
+
+	/**
+	 * Removes the specific mounting point for this modal from the DOM.
+	 */
+	cleanDOM() {
+		parentElement.removeChild( this.node );
+	}
+
+	/**
+	 * Appends the modal's node to the DOM, so the portal can render the
+	 * modal in it. Also calls the openFirstModal when this is the first modal to be
+	 * opened.
+	 */
+	componentDidMount() {
+		openModalCount++;
+
+		if ( openModalCount === 1 ) {
+			this.openFirstModal();
+		}
+	}
+
+	/**
+	 * Updates headingId when the aria.labelledby prop has changed.
+	 *
+	 * @param {Object} nextProps The component's next props.
+	 */
+	componentWillReceiveProps( nextProps ) {
```

Comment:
`componentWillReceiveProps` is effectively deprecated.

https://reactjs.org/docs/react-component.html#the-component-lifecycle

For this type of usage, you may consider `static getDerivedStateFromProps` instead:

https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops

However, in this case I don't think this requires state. I think we can just use the [`withInstanceId` higher-order component](https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-instance-id).

I will make these revisions in an upcoming commit.